File: Server-NestJS/scripts/verify-protocol-conformance.mjs
#!/usr/bin/env node

// SPDX-License-Identifier: Apache-2.0
/**
 * AI Governance Protocol Conformance (Moat 2.1 / A1)：协议合规认证套件。
 *
 * 独立、语言无关、确定性验证三大协议（docs/protocols/ai-governance-protocol.md）：
 *   1. 审计哈希链（§2）：canonicalJSON / hash / legacy 派生 / 链校验（篡改检测）。
 *   2. 委托 token（§3）：JWT HS256 签发 + 独立验签（aud 限定 / iss / exp / sub 前缀）。
 *   3. 工具风险分级（§4）：resolveRiskLevel 派生 + RISK_STRATEGY + requiresConfirmation。
 *   4. 治理绑定（§4.3/§4.4，CE-3 薄片）：策略→放行决策语义 + 授权拒绝依据词表。
 *
 * CE-1（2026-09-09）：断言数据源改为 specs/protocol/*-vector.json 机器可校验语料
 * （由 node scripts/generate-protocol-vectors.mjs 从现实现生成）——算法实现从
 * scripts/lib/protocol-algorithms.mjs 单源 import。金样本漂移由 generate --check 门禁拦截
 * （语义变更必须先落语料版本再改代码，CE-1 L3 单源规则）。
 *
 * 用法：node scripts/verify-protocol-conformance.mjs
 * 输出：docs/benchmark/protocol-conformance-<ts>.json + .md（机器可读报告）
 */
import { readFileSync, writeFileSync, mkdirSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import {
  canonicalJSON,
  legacyChainKey,
  chainHash,
  verifyChain,
  signJwt,
  verifyJwt,
  RISK_STRATEGY,
  resolveRiskLevel,
  needsConfirmation,
} from './lib/protocol-algorithms.mjs';

const __dirname = dirname(fileURLToPath(import.meta.url));
const SPECS = (name) => JSON.parse(readFileSync(resolve(__dirname, `../specs/protocol/${name}`), 'utf8'));

const startMs = Date.now();
const results = [];
const ok = (name, detail = '') => { results.push({ name, pass: true, detail }); console.log(`  ✓ ${name}${detail ? ` — ${detail}` : ''}`); };
const bad = (name, detail = '') => { results.push({ name, pass: false, detail }); console.log(`  ✗ ${name}${detail ? ` — ${detail}` : ''}`); };

console.log('═══ AI Governance Protocol Conformance（A1 · CE-1 语料驱动）═══\n');

/* ═══════════ 审计哈希链（协议 §2）——canonical 金样本 + hash 向量 ═══════════ */

console.log('─ 审计哈希链（协议 §2）─');

const canonicalVectors = SPECS('canonical-json-v1-vector.json');
for (const c of canonicalVectors.cases) {
  const got = canonicalJSON(c.input);
  got === c.canonicalBytes ? ok(c.name, got) : bad(c.name, `期望 ${c.canonicalBytes}，实得 ${got}`);
}

const hashVectors = SPECS('audit-hash-v1-vector.json');
for (const c of hashVectors.cases) {
  switch (c.kind) {
    case 'hex': {
      const got = chainHash(c.key, c.prevHash, c.payload);
      /^[0-9a-f]{64}$/.test(got) && got === c.expectHex
        ? ok(c.name, got.slice(0, 16) + '…')
        : bad(c.name, `期望 ${c.expectHex}，实得 ${got}`);
      break;
    }
    case 'determinism': {
      const h1 = chainHash(c.key, c.prevHash, c.payload);
      const h2 = chainHash(c.key, c.prevHash, c.payload);
      h1 === h2 && h1 === c.expectHex ? ok(c.name, '') : bad(c.name, `两次不一致或 ≠ 金样本（${h1}/${h2}）`);
      break;
    }
    case 'tamper': {
      const hA = chainHash(c.key, c.prevHash, c.payloadA);
      const hB = chainHash(c.key, c.prevHash, c.payloadB);
      hA !== hB ? ok(c.name, '') : bad(c.name, '篡改前后 hash 相同（异常）');
      break;
    }
    case 'genesis': {
      const genesisHash = chainHash(c.key, null, c.payload);
      const emptyStrHash = chainHash(c.key, '', c.payload);
      genesisHash !== emptyStrHash
        ? ok(c.name, '')
        : bad(c.name, 'genesis 与空串应产生不同 hash');
      break;
    }
    case 'legacy': {
      const got = legacyChainKey(c.secret);
      got === c.expectHex && /^[0-9a-f]{64}$/.test(got)
        ? ok(c.name, got.slice(0, 16) + '…')
        : bad(c.name, `期望 ${c.expectHex}，实得 ${got}`);
      break;
    }
    case 'chain': {
      const rows = c.rows;
      const keys = c.keys;
      const payloadFor = (row) => (c.payloadOverrides ?? []).find((o) => o.id === row.id)?.payload ?? row.payload;
      const got = verifyChain(rows, keys, payloadFor);
      const pass = c.expect.valid
        ? got.valid && got.checked === c.expect.checked
        : !got.valid && got.brokenIndex === c.expect.brokenIndex;
      pass
        ? ok(c.name, c.expect.valid ? `checked=${got.checked}` : `brokenIndex=${got.brokenIndex}`)
        : bad(c.name, JSON.stringify({ expect: c.expect, got }));
      break;
    }
    default:
      bad(c.name, `未知 kind：${c.kind}`);
  }
}

/* ═══════════ 委托 token（协议 §3）——claims 模板 + verifier 设置 ═══════════ */

console.log('\n─ 委托 token（协议 §3）─');

const tokenVectors = SPECS('delegation-token-v1-vector.json');
const baseNow = Math.floor(Date.now() / 1000);
for (const c of tokenVectors.cases) {
  const { iatBeforeSec = 60, expAfterSec = 240 } = c.lifetime ?? {};
  const claims = { ...c.claims, iat: baseNow - iatBeforeSec, exp: baseNow + expAfterSec };
  let token = signJwt(claims, c.secret);
  if (c.mutate === 'payload') {
    // 篡改 payload 段（保留原签名）→ 验签必须失败
    const [h, , sig] = token.split('.');
    token = `${h}.${Buffer.from('{"x":1}').toString('base64url')}.${sig}`;
  }
  const res = verifyJwt(token, c.secret, { audience: c.verifier?.audience, now: baseNow });
  let pass = res.ok === c.expect.ok;
  if (pass && c.expect.ok && c.expect.sub) pass = res.payload.sub === c.expect.sub;
  if (pass && c.expect.ok && c.expect.subPrefix) pass = res.payload.sub.startsWith(c.expect.subPrefix);
  if (pass && !c.expect.ok && res.reason) {
    ok(c.name, res.reason);
  } else if (pass) {
    ok(c.name, c.expect.ok ? `sub=${res.payload.sub}` : '');
  } else {
    bad(c.name, `期望 ${JSON.stringify(c.expect)}，实得 ${JSON.stringify(res)}`);
  }
}

/* ═══════════ 工具风险分级（协议 §4） ═══════════ */

console.log('\n─ 工具风险分级（协议 §4）─');

const riskVectors = SPECS('risk-level-v1-vector.json');
const strategiesMatch = JSON.stringify(RISK_STRATEGY) === JSON.stringify(riskVectors.strategies);
strategiesMatch ? ok('RISK_STRATEGY 表与语料一致', '') : bad('RISK_STRATEGY 表与语料一致', '实现与语料 strategies 漂移');
for (const c of riskVectors.cases) {
  const level = resolveRiskLevel(c.in);
  const pass = level === c.expect.level && RISK_STRATEGY[level] === c.expect.strategy && needsConfirmation(level) === c.expect.needsConfirmation;
  pass
    ? ok(c.name, `${level}/${RISK_STRATEGY[level]}`)
    : bad(c.name, `期望 ${JSON.stringify(c.expect)}，实得 ${level}/${RISK_STRATEGY[level]}`);
}

/* ═══════════ 治理绑定（协议 §4.3/§4.4，CE-3 薄片） ═══════════ */

console.log('\n─ 治理绑定（协议 §4.3/§4.4）─');

const bindingVectors = SPECS('governance-binding-v1-vector.json');
{
  const strategies = [...new Set(Object.values(RISK_STRATEGY))].sort();
  const bound = Object.keys(bindingVectors.gateOutcomeByStrategy).sort();
  JSON.stringify(bound) === JSON.stringify(strategies)
    ? ok('gateOutcomeByStrategy 覆盖 RISK_STRATEGY 全部策略值', bound.join(','))
    : bad('gateOutcomeByStrategy 覆盖 RISK_STRATEGY 全部策略值', `期望 ${strategies}，实得 ${bound}`);

  let allOk = true;
  for (const d of bindingVectors.derivation) {
    const o = bindingVectors.gateOutcomeByStrategy[d.strategy];
    const consistent =
      !!o &&
      o.requiresConfirmation === d.requiresConfirmation &&
      (o.requiresConfirmation || o.requiresApproval) === needsConfirmation(d.riskLevel) &&
      o.blocked === (d.strategy === 'block') &&
      o.executes === !(o.requiresConfirmation || o.requiresApproval || o.blocked);
    if (!consistent) {
      allOk = false;
      bad(`绑定一致 ${d.riskLevel}`, JSON.stringify({ derivation: d, outcome: o }));
    }
  }
  allOk &&
    ok(
      'R0..R5 → 策略 → 放行决策 逐级一致（executes/确认/审批/阻断 与 needsConfirmation 互洽）',
      `${bindingVectors.derivation.length} 级`,
    );

  const denies = bindingVectors.denyChecks;
  denies.length > 0 && new Set(denies).size === denies.length
    ? ok('授权拒绝依据词表唯一且非空（可解释授权封闭词汇）', denies.join(','))
    : bad('授权拒绝依据词表唯一且非空', JSON.stringify(denies));
}

/* ═══════════ 报告 ═══════════ */

const passCount = results.filter((r) => r.pass).length;
const elapsed = Date.now() - startMs;
const ts = new Date().toISOString().replace(/[:.]/g, '-');
const report = {
  gate: 'AI Governance Protocol Conformance（协议合规认证套件，A1 · CE-1 语料驱动）',
  protocol: 'ai-governance-protocol（审计链 / 委托 token / 工具风险分级）',
  vectors: {
    canonical: canonicalVectors.vectorVersion,
    auditHash: hashVectors.vectorVersion,
    delegationToken: tokenVectors.vectorVersion,
    riskLevel: riskVectors.vectorVersion,
    governanceBinding: bindingVectors.vectorVersion,
  },
  date: ts,
  pass: passCount,
  total: results.length,
  elapsedSec: Math.round(elapsed / 1000),
  implementations: [
    {
      name: 'Server-NestJS（参考实现，语料生成源）',
      audited: true,
      auditedAt: ts,
    },
  ],
  cases: results,
};
mkdirSync(resolve(__dirname, '../docs/benchmark'), { recursive: true });
const base = resolve(__dirname, `../docs/benchmark/protocol-conformance-${ts}`);
writeFileSync(`${base}.json`, JSON.stringify(report, null, 2));
const md = [
  `# AI Governance Protocol Conformance（${ts}）`, '',
  `- ${passCount}/${results.length} 通过 ｜ 总耗时 ${Math.round(elapsed / 1000)}s ｜ 协议：审计链 / 委托 token / 工具风险分级（语料驱动，specs/protocol）`, '',
  '| # | 断言 | 结果 | 详情 |', '|---|------|------|------|',
  ...results.map((r, i) => `| ${i + 1} | ${r.name} | ${r.pass ? '✅' : '❌'} | ${r.detail} |`), '',
].join('\n');
writeFileSync(`${base}.md`, md);

console.log(`\n═══ Conformance 结果：${passCount}/${results.length} 通过（${Math.round(elapsed / 1000)}s）═══`);
console.log(`报告：docs/benchmark/protocol-conformance-${ts}.md`);
process.exit(passCount === results.length ? 0 : 1);

File: Server-NestJS/scripts/verify-protocol-conformance.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
"""语言中性合规 runner（协议 §2 审计链）——**第二语言实现**。

证明 `specs/protocol/*-vector.json`（JSON 语料）可被**非 Node** 实现消费：本脚本**独立实现**
协议 §2.2/§2.3（canonicalJSON + 链 hash + legacy 派生 + 链校验），**不 import 任何 KeelBase 源码**，
对照金样本复算即自证一致（见 docs/protocols/conformance-profile.md §2.1 Core）。

用法：python Server-NestJS/scripts/verify-protocol-conformance.py
"""
import hashlib
import hmac
import json
import os
import sys

# Windows 控制台默认 GBK，强制 UTF-8 以正确输出中文/符号（不影响算法）
try:
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
except Exception:  # noqa: BLE001 —— 老环境无 reconfigure 时降级
    pass

SPECS = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "specs", "protocol")

# ── §2.3 canonicalJSON（语言中性规则：K 排序 + 每层键白名单 + ECMAScript 数字/JSON 转义）──

def _utf16(s: str) -> bytes:
    """键排序按 UTF-16 code unit（Java String.compareTo / JS 默认 sort 同序）。"""
    return s.encode("utf-16-be")


def js_number(x) -> str:
    """数字 → ECMAScript Number::toString（§2.3 规则 5）：短往返、-0→0、|x|>=1e21 或 <1e-6 用指数。"""
    if isinstance(x, bool):
        raise TypeError("bool 不是 number")
    if isinstance(x, int):
        return str(x)
    f = float(x)
    if f == 0:
        return "0"  # -0.0 → 0
    sign = "-" if f < 0 else ""
    f = abs(f)
    r = repr(f)  # Python repr = 最短往返（与 JS 同源算法）
    if "e" in r:
        mant, exp = r.split("e")
        e = int(exp)
    else:
        mant, e = r, 0
    ip, _, fp = mant.partition(".")
    full = ip + fp
    digits = full.lstrip("0")
    if digits == "":
        return "0"
    lead = len(full) - len(digits)
    n = e + len(ip) - lead  # 值 = digits × 10^(n - k)，k = len(digits)
    k = len(digits)
    if k <= n <= 21:
        return sign + digits + "0" * (n - k)
    if 0 < n <= 21:
        return sign + digits[:n] + "." + digits[n:]
    if -6 < n <= 0:
        return sign + "0." + "0" * (-n) + digits
    ex = n - 1
    mant_out = digits[0] + ("." + digits[1:] if k > 1 else "")
    return sign + mant_out + "e" + ("+" if ex >= 0 else "-") + str(abs(ex))


def _ser(v, kset) -> str:
    if v is None:
        return "null"
    if v is True:
        return "true"
    if v is False:
        return "false"
    if isinstance(v, (int, float)):
        return js_number(v)
    if isinstance(v, str):
        return json.dumps(v, ensure_ascii=False)  # §2.3 规则 4：仅必需转义、非 ASCII 不转
    if isinstance(v, list):
        return "[" + ",".join(_ser(e, kset) for e in v) + "]"
    if isinstance(v, dict):
        # §2.3 规则 1：每层只输出 K 中存在的键，按 K 序（对象自身键序无关）
        return "{" + ",".join(
            json.dumps(k, ensure_ascii=False) + ":" + _ser(v[k], kset) for k in kset if k in v
        ) + "}"
    raise TypeError(f"不可序列化：{type(v)}")


def canonical_json(payload: dict) -> str:
    # JSON 语料无 undefined；K = 顶层键集（剔除 undefined 的语义在 JSON 下为空操作）
    kset = sorted(payload.keys(), key=_utf16)
    return _ser(payload, kset)


def legacy_chain_key(secret: str) -> str:
    return hmac.new(b"keelbase:audit-chain:v1", secret.encode("utf-8"), hashlib.sha256).hexdigest()


def chain_hash(key: str, prev_hash, payload: dict) -> str:
    msg = f"{prev_hash if prev_hash is not None else 'genesis'}|{canonical_json(payload)}"
    return hmac.new(key.encode("utf-8"), msg.encode("utf-8"), hashlib.sha256).hexdigest()


def verify_chain(rows, keys, payload_for) -> dict:
    prev = None
    for i, row in enumerate(rows):
        if row.get("prevHash") != prev or not any(
            chain_hash(k, prev, payload_for(row)) == row.get("hash") for k in keys
        ):
            return {"valid": False, "checked": i, "brokenIndex": i + 1}
        prev = row.get("hash")
    return {"valid": True, "checked": len(rows)}


# ── runner ──

CASES = []


def ok(name, detail=""):
    CASES.append(True)
    print(f"  ✓ {name}" + (f" — {detail}" if detail else ""))


def bad(name, detail=""):
    CASES.append(False)
    print(f"  ✗ {name}" + (f" — {detail}" if detail else ""))


def load(name):
    with open(os.path.join(SPECS, name), encoding="utf-8") as fh:
        return json.load(fh)


def main() -> int:
    print("═══ 语言中性合规 runner（Python · 协议 §2 审计链）═══")

    print("─ canonicalJSON（§2.3）─")
    for c in load("canonical-json-v1-vector.json")["cases"]:
        got = canonical_json(c["input"])
        ok(c["name"], got) if got == c["canonicalBytes"] else bad(c["name"], f"期望 {c['canonicalBytes']}，实得 {got}")

    print("─ 审计哈希链（§2.2）─")
    for c in load("audit-hash-v1-vector.json")["cases"]:
        kind = c["kind"]
        if kind == "hex":
            got = chain_hash(c["key"], c["prevHash"], c["payload"])
            ok(c["name"], got[:16] + "…") if got == c["expectHex"] else bad(c["name"], f"实得 {got}")
        elif kind == "determinism":
            h1 = chain_hash(c["key"], c["prevHash"], c["payload"])
            h2 = chain_hash(c["key"], c["prevHash"], c["payload"])
            ok(c["name"]) if h1 == h2 and h1 == c["expectHex"] else bad(c["name"], "不一致或 ≠ 金样本")
        elif kind == "tamper":
            hA = chain_hash(c["key"], c["prevHash"], c["payloadA"])
            hB = chain_hash(c["key"], c["prevHash"], c["payloadB"])
            ok(c["name"]) if hA != hB else bad(c["name"], "篡改前后 hash 相同")
        elif kind == "genesis":
            g = chain_hash(c["key"], None, c["payload"])
            e = chain_hash(c["key"], "", c["payload"])
            ok(c["name"]) if g != e else bad(c["name"], "genesis 与空串应不同")
        elif kind == "legacy":
            got = legacy_chain_key(c["secret"])
            ok(c["name"], got[:16] + "…") if got == c["expectHex"] else bad(c["name"], f"实得 {got}")
        elif kind == "chain":
            overrides = {o["id"]: o["payload"] for o in c.get("payloadOverrides", [])}
            payload_for = lambda r: overrides.get(r["id"], r["payload"])
            got = verify_chain(c["rows"], c["keys"], payload_for)
            exp = c["expect"]
            passed = (
                got["valid"] == exp["valid"]
                and (got.get("checked") == exp.get("checked") if exp["valid"] else got.get("brokenIndex") == exp.get("brokenIndex"))
            )
            ok(c["name"], f"checked={got.get('checked')}" if exp["valid"] else f"brokenIndex={got.get('brokenIndex')}") if passed else bad(c["name"], json.dumps({"expect": exp, "got": got}))
        else:
            bad(c["name"], f"未知 kind：{kind}")

    passed, total = sum(CASES), len(CASES)
    print(f"\n═══ 结果：{passed}/{total} 通过 ═══")
    return 0 if passed == total else 1


if __name__ == "__main__":
    sys.exit(main())
